### ISSUE TITLE
Wrong poses due to duplicate/symmetric features

### ISSUE THREAD

**Issue by Athrunen** [colmap/colmap#1437]
I'm trying to reconstruct a room's interior from video game screenshots. Which works surprisingly well.

 But the fact that a lot of the background features are pixel-perfect replicas of one another and that the room is symmetrical creates (at least) two versions of the same room flipped and superimposed over one another.

![Overlap](https://user-images.githubusercontent.com/13453494/155373599-69ce6c93-c141-487c-9a5e-46c5f3b81afd.png)

How would I go about fixing this? 

I've already tried the adjustments in #1215 but they had no discernable effects.
Masking could help with the background features but that would also remove them from the reconstruction as a whole. 
Plus it's a bit unfeasible to mask over 700 images when there could be better options. That is also the reason I have not yet manually matched the images.

**Comment by tsattler** [comment]
I don't think that there is an easy fix for this. Rather, you would probably need to modify source code, e.g., you could try to look at the distribution of matching features in the images to draw conclusions on whether this is the same room or not.

Colmap does not implement mechanisms to handle ambiguous matches. You could also have a look at https://github.com/cvg/sfm-disambiguation-colmap 
But I think the conclusion of that work is that there currently is no universal method to handle ambiguities.

**Comment by Athrunen** [comment]
It might be easier to more or less brute-force a solution than to filter out incorrect matches beforehand. Especially in a big set where there are at least some full paths around the object.

Because then we have very similar poses somewhere in the set that are a much closer match than the wrong matches.
That means after we have mapped all the images we can try to remove and remap some of them to where their overall error is less than before.

And the simplest of those approaches would be to just pick some percent of the images at random, remove and try to fit them again and after some iteration compare it to the average error before and either keep or reject the changes.

Does that sound reasonable?

**Comment by tsattler** [comment]
I am not sure I understand.

**Comment by Athrunen** [comment]
What part do you not understand?

**Comment by tsattler** [comment]
I understand the basic idea, but I am not sure how you would measure the overall error when you say "we can try to remove and remap some of them to where their overall error is less than before". 

I don't see how randomly removing images will help. Why would there be a better place to reconnect them than their current place? You would need to identify images that are connecting unrelated parts (e.g., using something akin to https://www.cs.unc.edu/~jheinly/duplicate_structure.html that reasons about missing observations), obtain individual connected components, and then try to stitch these components together again.

**Comment by Athrunen** [comment]
I have to admit I don't know much about either the actual implementation detail or the theory behind most of this stuff.

But from reading the source code I'm making the following assumptions:

1. That if I move an image to a correct pose the models overall (reprojection) error will drop.
2. That an image introduced to a model that maps the overall shape(i.e not just half the model) correctly, but has not necessarily all poses in the right places, tends towards the pose where that image has the most matches.
3. Given a high enough amount of images used for such a model, the pose with the most matches for a specific image approaches the "correct" pose for the image. (Up to some threshold of wrong poses)
4. To combat such a threshold, the more wrong poses there are the more images we have to remove at the same time to not converge to old poses.

Is that correct?

**Comment by tsattler** [comment]
1. This is not necessarily correct: at the true pose, you might have more 2D-3D correspondences than at the wrong pose. At the same time, you already have quite low reprojection errors for the wrong pose. There thus is a good chance that the sum of reprojection errors is larger.
2. It really depends what is currently in your map (remember that Colmap builds maps by adding a few images at a time to an initial map). If the map only contains the repeating structure and an image has many matches with this structure, it will be connected there.
3. Simple counter example: consider a building facade with many windows. The first image shows the first three windows. The second image shows windows 2, 3, and 4, the third image shows windows 3, 4, and 5, etc. You will find the most matches by aligning each pair of images in such a way that all three windows shown in it are matched.
4. I don't see why this would be the case. A few wrongly placed images might be all that might need to be removed.

**Comment by Athrunen** [comment]
1. So you mean that because we have more matching points, we have more points that could be projected wrongly, thus actually increasing the error?
2. Ok, I was basing this somewhat on my case where the whole room is mapped but some features are duplicated or overlapping.
3. Is that case even solvable? If we have no differentiating features around these windows, even humans that know the building will have a hard time placing the images in relation to the others. Only if the images are in sequence this is even remotely approachable.
4. Yes, but the more wrongly placed images we have the more chances they have to cluster and attract more wrong images. Those clusters are in themselves stable and only removing a few wrong poses will let them return to their old state.

**Comment by tsattler** [comment]
1. All points will have a slight error as you have errors in your 2D feature detections. The more 2D-3D matches you have, the larger will be your sum of reprojection errors, even if there are no wrong 3D points.
3. Probably not without some external knowledge.

**Comment by Athrunen** [comment]
The alternative would be to keep track of the number of feature matches before and after redistributing images and use those to determine if the image is placed correctly. But then we disregard the reprojection error...

I'm already halfway done with implementing the rough removing and adding images part in a local fork. Gonna try out some different combinations of factors when that is done.

Is there any other existing research that might help me out with that, beyond the repo already pointed out?

As for my current project, I've used a video of the scene in question with a very tight sequential overlap and no loop detection and that certainly helped, but tight turns with the camera will sometimes cause the scene to misalign at a weird angle.

**Comment by tsattler** [comment]
The alternative sounds a bit like the work by Heinly et al.

You can follow-up on the research by Heinly et al. and the pointers provided in the first repository.

**Comment by snavely** [comment]
You could also try out this newer method, if you wanted to! https://doppelgangers-3d.github.io/
